include/envoy/stats/symbol_table.h

```diff
@@ -1,13 +1,157 @@
 #pragma once
 
+#include <memory>
+
+#include "envoy/common/pure.h"
+
+#include "absl/strings/string_view.h"
+
 namespace Envoy {
 namespace Stats {
 
-// Interface for referencing a stat name.
 class StatName;
+class SymbolEncoding;
+
+/**
+ * Efficient byte-encoded storage of an array of tokens. The most common tokens
+ * are typically < 127, and are represented directly. tokens >= 128 spill into
+ * the next byte, allowing for tokens of arbitrary numeric value to be stored.
+ * As long as the most common tokens are low-valued, the representation is
+ * space-efficient. This scheme is similar to UTF-8.
+ */
+using SymbolStorage = uint8_t[];
+using SymbolStoragePtr = std::unique_ptr<SymbolStorage>;
+
+/**
+ * SymbolTable manages a namespace optimized for stats, which are typically
+ * composed of arrays of "."-separated tokens, with a significant overlap
+ * between the tokens. Each token is mapped to a Symbol (uint32_t) and
+ * reference-counted so that no-longer-used symbols can be reclaimed.
+ *
+ * We use a uint8_t array to encode arrays of symbols in order to conserve
+ * space, as in practice the majority of token instances in stat names draw from
+ * a fairly small set of common names, typically less than 100. The format is
+ * somewhat similar to UTF-8, with a variable-length array of uint8_t. See the
+ * implementation for details.
```

Comment:
Abstract headers should standalone, I don't think it should be necessary to draw on implementation details here.